src/programs/azimuthal_average/azimuthal_average.cpp

```diff
@@ -0,0 +1,1534 @@
+	sum_image.QuickAndDirtyWriteSlice("my_rotationally_averaged_sum.mrc", 1);
+
+	// get scale factor to multiply reference with
+	wxPrintf("\nScaling Reference...\n");
+	scale_reference(image_stack, number_of_input_images, &my_volume, ctf_parameters, phase_flip_only, max_threads);
+
+	// print out 3D reconstruction
+	my_volume.QuickAndDirtyWriteSlices("my_averaged_volume.mrc", 1, my_volume.logical_z_dimension);
+
+	// save orthogonal views
+	Image orth_image;
+	orth_image.Allocate(my_volume.logical_x_dimension * 3, my_volume.logical_y_dimension * 2, 1, true);
+	my_volume.CreateOrthogonalProjectionsImage(&orth_image);
+	orth_image.QuickAndDirtyWriteSlice("my_orthogonal_views.mrc", 1);	
+
+	// clean-up
+	if (input_ctf_values_from_text_file == true) delete input_text;
+	//delete my_progress;
+	delete[] x_shifts;
+	delete[] y_shifts;
+	delete[] psi_angles;
+	delete[] image_stack;
+	delete[] image_stack_times_ctf;
+	delete[] ctf_sum_of_squares;
+	for (int i = 0; i < number_of_input_images; i++)
+	{
+		delete[] ctf_parameters[i];
+	}
+	delete[] ctf_parameters;
+	wxPrintf("\n\n");
+	
+	overall_finish = wxDateTime::Now();
+
+	// Timings
+	wxPrintf("\nTimings\n");
+	wxTimeSpan time_total = overall_finish.Subtract(overall_start);
+	wxPrintf(" Overal time                : %s\n",time_total.Format());
+	
+	return true;
+}
+
+void apply_ctf(Image *current_image, CTF ctf_to_apply, float *ctf_sum_of_squares, bool absolute)
+{
+	float y_coord_sq;
+	float x_coord_sq;
+
+	float y_coord;
+	float x_coord;
+
+	float frequency_squared;
+	float azimuth;
+	float ctf_value;
+
+	long pixel_counter = 0;
+
+	for (int j = 0; j <= current_image->physical_upper_bound_complex_y; j++)
+	{
+		y_coord = current_image->ReturnFourierLogicalCoordGivenPhysicalCoord_Y(j) * current_image->fourier_voxel_size_y;
+		y_coord_sq = powf(y_coord, 2.0);
+
+		for (int i = 0; i <= current_image->physical_upper_bound_complex_x; i++)
+		{
+			x_coord = i * current_image->fourier_voxel_size_x;
+			x_coord_sq = powf(x_coord, 2.0);
+
+			// Compute the azimuth
+			if (i == 0 && j == 0)
+			{
+				azimuth = 0.0;
+			}
+			else
+			{
+				azimuth = atan2f(y_coord, x_coord);
+			}
+
+			// Compute the square of the frequency
+			frequency_squared = x_coord_sq + y_coord_sq;
+			ctf_value = ctf_to_apply.Evaluate(frequency_squared, azimuth);
+
+			// phase-flip
+			if (absolute) ctf_value = fabsf(ctf_value);
+
+			current_image->complex_values[pixel_counter] *= ctf_value;
+			ctf_sum_of_squares[pixel_counter] += powf(ctf_value, 2);
+			pixel_counter++;
+		}
+	}
+}
+
+void divide_by_ctf_sum_of_squares(Image *current_image, float *ctf_sum_of_squares)
+{
+	// normalize by sum of squared CTFs (voxel by voxel)
+	long pixel_counter = 0;
+
+	for (int j = 0; j <= current_image->physical_upper_bound_complex_y; j++)
+	{
+		for (int i = 0; i <= current_image->physical_upper_bound_complex_x; i++)
+		{
+			current_image->complex_values[pixel_counter] /= sqrtf(ctf_sum_of_squares[pixel_counter]);
```

Comment:
Is this definitely non-zero division?

Reply:
To avoid division by zero, I can add an if statement?

`if (ctf_sum_of_squares[pixel_counter] != 0.0) current_image->complex_values[pixel_counter] /= sqrtf(ctf_sum_of_squares[pixel_counter]);`

Reply:
To add, I guess a Wiener filter would add a noise-to-signal constant in the denominator to possibly dividing by zero or a small number.

Reply:
I would think the latter option is preferable, but check with Tim on that. What if it is 1e-6? You pass your conditional statement, and avoid an inf, but still have a super large number.